storage/innobase/arch/arch0page.cc

```diff
@@ -0,0 +1,3070 @@
+
+  if (start_id != 0 && stop_id != 0 && start_id >= stop_id)
+    return ER_PAGE_TRACKING_RANGE_NOT_TRACKED;
+
+  arch_oper_mutex_enter();
+  auto latest_stop_lsn= m_latest_stop_lsn;
+  arch_oper_mutex_exit();
+
+  ut_ad(latest_stop_lsn != LSN_MAX);
+
+  if (start_id == 0 || stop_id == 0)
+  {
+    if (m_current_group == nullptr || !m_current_group->is_active())
+      return ER_PAGE_TRACKING_RANGE_NOT_TRACKED;
+
+    *group= m_current_group;
+    ut_ad(m_last_lsn != LSN_MAX);
+
+    start_id= (start_id == 0) ? m_last_lsn : start_id;
+    stop_id= (stop_id == 0) ? latest_stop_lsn : stop_id;
+  }
+
+  if (start_id >= stop_id || start_id == LSN_MAX || stop_id == LSN_MAX)
+    return ER_PAGE_TRACKING_RANGE_NOT_TRACKED;
+
+  if (*group == nullptr)
+  {
+    for (auto it : m_group_list)
+    {
+      *group= it;
+      if (start_id < (*group)->get_begin_lsn() ||
+          (!(*group)->is_active() && stop_id > (*group)->get_end_lsn()) ||
+          ((*group)->is_active() && stop_id > latest_stop_lsn))
+      {
+        *group= nullptr;
+        continue;
+      }
+      break;
+    }
+  }
+  return (*group) ? 0 : ER_PAGE_TRACKING_RANGE_NOT_TRACKED;
+}
+
+uint Arch_Page_Sys::purge(lsn_t *purge_lsn)
+{
+  lsn_t purged_lsn= LSN_MAX;
+  uint err= 0;
+
+  if (*purge_lsn == 0)
+    *purge_lsn = log_sys.last_checkpoint_lsn.load(std::memory_order_seq_cst);
+
+  DBUG_PRINT("page_archiver", ("Purging of files - %" PRIu64 "", *purge_lsn));
+  arch_mutex_enter();
+
+  for (auto it = m_group_list.begin(); it != m_group_list.end();)
+  {
+    lsn_t group_purged_lsn= LSN_MAX;
+    auto group= *it;
+    DBUG_PRINT("page_archiver",
+               ("End lsn - %" PRIu64 "", group->get_end_lsn()));
+
+    err= group->purge(*purge_lsn, group_purged_lsn);
+
+    if (group_purged_lsn == LSN_MAX)
+      break;
+
+    DBUG_PRINT("page_archiver",
+               ("Group purged lsn - %" PRIu64 "", group_purged_lsn));
+
+    if (purged_lsn == LSN_MAX || group_purged_lsn > purged_lsn)
+      purged_lsn= group_purged_lsn;
+
+    if (!group->is_active() && group->get_end_lsn() <= group_purged_lsn)
+    {
+      it= m_group_list.erase(it);
+      UT_DELETE(group);
+
+      DBUG_PRINT("page_archiver", ("Purged entire group."));
+      continue;
+    }
+    ++it;
+  }
+  DBUG_PRINT("page_archiver",
+             ("Purged archived file until : %" PRIu64 "", purged_lsn));
+  *purge_lsn= purged_lsn;
+
+  if (purged_lsn == LSN_MAX)
+  {
+    arch_mutex_exit();
+    return err;
+  }
+  m_latest_purged_lsn = purged_lsn;
+  arch_mutex_exit();
+
+  return err;
+}
+
+void Arch_Page_Sys::update_stop_info(Arch_Block *cur_blk)
+{
+  mysql_mutex_assert_owner(&m_oper_mutex);
+
+  if (cur_blk != nullptr)
+    cur_blk->update_block_header(m_latest_stop_lsn, LSN_MAX);
+
+  if (m_current_group != nullptr)
+    m_current_group->update_stop_point(m_write_pos, m_latest_stop_lsn);
+}
+
+#ifdef UNIV_DEBUG
+void Arch_Page_Sys::print()
+{
+  DBUG_PRINT("page_archiver", ("State : %u", m_state));
```

Comment:
Unreachable function. And yes, I was testing `cmake -DCMAKE_BUILD_TYPE=DEBUG -DENABLE_GCOV=ON`.